sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/ProactiveConnectionManagementTest.java

```diff
@@ -0,0 +1,393 @@
+import com.azure.cosmos.implementation.routing.PartitionKeyInternalHelper;
+import com.azure.cosmos.implementation.routing.PartitionKeyRangeIdentity;
+import com.azure.cosmos.models.CosmosContainerIdentity;
+import com.azure.cosmos.rx.TestSuiteBase;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.azure.cosmos.implementation.TestUtils.mockDiagnosticsClientContext;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+public class ProactiveConnectionManagementTest extends TestSuiteBase {
+
+    private CosmosClientBuilder clientBuilder;
+    private DatabaseAccount databaseAccount;
+    private CosmosAsyncDatabase cosmosAsyncDatabase;
+
+    @BeforeClass(groups = {"multi-region"})
+    public void beforeClass() {
+        clientBuilder = new CosmosClientBuilder()
+                .endpoint(TestConfigurations.HOST)
+                .key(TestConfigurations.MASTER_KEY)
+                .contentResponseOnWriteEnabled(true)
+                .directMode();
+
+        CosmosAsyncClient dummyClient = new CosmosClientBuilder()
+                .endpoint(TestConfigurations.HOST)
+                .key(TestConfigurations.MASTER_KEY)
+                .contentResponseOnWriteEnabled(true)
+                .directMode().buildAsyncClient();
+
+        this.cosmosAsyncDatabase = getSharedCosmosDatabase(dummyClient);
+
+        AsyncDocumentClient asyncDocumentClient = ReflectionUtils.getAsyncDocumentClient(dummyClient);
+        RxDocumentClientImpl rxDocumentClient = (RxDocumentClientImpl) asyncDocumentClient;
+        GlobalEndpointManager globalEndpointManager =
+                ReflectionUtils.getGlobalEndpointManager(rxDocumentClient);
+        this.databaseAccount = globalEndpointManager.getLatestDatabaseAccount();
+
+        safeClose(dummyClient);
+    }
+
+    @Test(groups = {"multi-region"}, dataProvider = "invalidProactiveContainerInitConfigs")
+    public void openConnectionsAndInitCachesWithInvalidCosmosClientConfig(List<String> preferredRegions, int numProactiveConnectionRegions, int numContainers) {
+
+        CosmosAsyncClient clientWithOpenConnections = null;
+
+        List<CosmosAsyncContainer> asyncContainers = new ArrayList<>();
+        List<CosmosContainerIdentity> cosmosContainerIdentities = new ArrayList<>();
+
+        for (int i = 1; i <= numContainers; i++) {
+            String containerId = String.format("id%d", i);
+            cosmosAsyncDatabase.createContainerIfNotExists(containerId, "/mypk").block();
```

Comment:
for this test, creating those containers seems unnecessary

Reply:
Won't fix - no harm.